File: gamification.py
# Gamification system for Pomodoro Timer
import json
import os
from datetime import datetime, date, timedelta
from typing import Dict, List, Any
from dataclasses import dataclass, asdict
from enum import Enum

class AchievementType(Enum):
    FIRST_COMPLETION = "first_completion"
    DAILY_STREAK_3 = "daily_streak_3"
    DAILY_STREAK_7 = "daily_streak_7"
    WEEKLY_10 = "weekly_10"
    WEEKLY_20 = "weekly_20"
    TOTAL_50 = "total_50"
    TOTAL_100 = "total_100"

@dataclass
class Achievement:
    id: str
    name: str
    description: str
    icon: str
    earned: bool = False
    earned_date: str = None

@dataclass
class UserStats:
    total_xp: int = 0
    level: int = 1
    total_sessions: int = 0
    current_streak: int = 0
    best_streak: int = 0
    last_session_date: str = None
    achievements: List[Dict] = None
    weekly_sessions: Dict[str, int] = None
    monthly_sessions: Dict[str, int] = None

    def __post_init__(self):
        if self.achievements is None:
            self.achievements = []
        if self.weekly_sessions is None:
            self.weekly_sessions = {}
        if self.monthly_sessions is None:
            self.monthly_sessions = {}

class GamificationManager:
    def __init__(self, data_file='data/gamification.json'):
        self.data_file = data_file
        self.achievements_definitions = {
            AchievementType.FIRST_COMPLETION: Achievement(
                "first_completion", "First Steps", "Complete your first Pomodoro session", "🍅"
            ),
            AchievementType.DAILY_STREAK_3: Achievement(
                "daily_streak_3", "Consistent", "Complete sessions for 3 consecutive days", "🔥"
            ),
            AchievementType.DAILY_STREAK_7: Achievement(
                "daily_streak_7", "Dedicated", "Complete sessions for 7 consecutive days", "⚡"
            ),
            AchievementType.WEEKLY_10: Achievement(
                "weekly_10", "Weekly Warrior", "Complete 10 sessions in a week", "💪"
            ),
            AchievementType.WEEKLY_20: Achievement(
                "weekly_20", "Focus Master", "Complete 20 sessions in a week", "👑"
            ),
            AchievementType.TOTAL_50: Achievement(
                "total_50", "Half Century", "Complete 50 total sessions", "🏆"
            ),
            AchievementType.TOTAL_100: Achievement(
                "total_100", "Centurion", "Complete 100 total sessions", "🌟"
            )
        }
        self.user_stats = self._load_stats()

    def _load_stats(self) -> UserStats:
        """Load user statistics from file"""
        if os.path.exists(self.data_file):
            try:
                with open(self.data_file, 'r') as f:
                    data = json.load(f)
                return UserStats(**data)
            except (json.JSONDecodeError, TypeError):
                pass
        return UserStats()

    def _save_stats(self):
        """Save user statistics to file"""
        os.makedirs(os.path.dirname(self.data_file), exist_ok=True)
        with open(self.data_file, 'w') as f:
            json.dump(asdict(self.user_stats), f, indent=2)

    def _calculate_level(self, xp: int) -> int:
        """Calculate level based on XP (100 XP per level)"""
        return max(1, (xp // 100) + 1)

    def _get_week_key(self, date_obj: date = None) -> str:
        """Get week key in format YYYY-WW"""
        if date_obj is None:
            date_obj = date.today()
        year, week, _ = date_obj.isocalendar()
        return f"{year}-{week:02d}"

    def _get_month_key(self, date_obj: date = None) -> str:
        """Get month key in format YYYY-MM"""
        if date_obj is None:
            date_obj = date.today()
        return f"{date_obj.year}-{date_obj.month:02d}"

    def complete_session(self) -> Dict[str, Any]:
        """Handle session completion and return updated stats with new achievements"""
        today = date.today().isoformat()
        xp_gained = 25  # 25 XP per completed session
        
        # Update basic stats
        self.user_stats.total_xp += xp_gained
        self.user_stats.total_sessions += 1
        old_level = self.user_stats.level
        self.user_stats.level = self._calculate_level(self.user_stats.total_xp)
        
        # Update streak
        yesterday = (date.today() - timedelta(days=1)).isoformat()
        if self.user_stats.last_session_date == yesterday:
            self.user_stats.current_streak += 1
        elif self.user_stats.last_session_date != today:  # New streak if not today
            self.user_stats.current_streak = 1
        
        self.user_stats.best_streak = max(self.user_stats.best_streak, self.user_stats.current_streak)
        self.user_stats.last_session_date = today
        
        # Update weekly/monthly stats
        week_key = self._get_week_key()
        month_key = self._get_month_key()
        self.user_stats.weekly_sessions[week_key] = self.user_stats.weekly_sessions.get(week_key, 0) + 1
        self.user_stats.monthly_sessions[month_key] = self.user_stats.monthly_sessions.get(month_key, 0) + 1
        
        # Check for new achievements
        new_achievements = self._check_achievements()
        
        # Save updated stats
        self._save_stats()
        
        return {
            'xp_gained': xp_gained,
            'total_xp': self.user_stats.total_xp,
            'level': self.user_stats.level,
            'level_up': self.user_stats.level > old_level,
            'current_streak': self.user_stats.current_streak,
            'new_achievements': new_achievements,
            'total_sessions': self.user_stats.total_sessions
        }

    def _check_achievements(self) -> List[Dict]:
        """Check and award new achievements"""
        new_achievements = []
        earned_ids = {ach['id'] for ach in self.user_stats.achievements}
        
        # Check each achievement type
        checks = [
            (AchievementType.FIRST_COMPLETION, self.user_stats.total_sessions >= 1),
            (AchievementType.DAILY_STREAK_3, self.user_stats.current_streak >= 3),
            (AchievementType.DAILY_STREAK_7, self.user_stats.current_streak >= 7),
            (AchievementType.WEEKLY_10, max(self.user_stats.weekly_sessions.values(), default=0) >= 10),
            (AchievementType.WEEKLY_20, max(self.user_stats.weekly_sessions.values(), default=0) >= 20),
            (AchievementType.TOTAL_50, self.user_stats.total_sessions >= 50),
            (AchievementType.TOTAL_100, self.user_stats.total_sessions >= 100),
        ]
        
        for achievement_type, condition in checks:
            achievement_def = self.achievements_definitions[achievement_type]
            if condition and achievement_def.id not in earned_ids:
                earned_achievement = asdict(achievement_def)
                earned_achievement['earned'] = True
                earned_achievement['earned_date'] = date.today().isoformat()
                self.user_stats.achievements.append(earned_achievement)
                new_achievements.append(earned_achievement)
        
        return new_achievements

    def get_stats(self) -> Dict[str, Any]:
        """Get current user statistics"""
        return {
            'total_xp': self.user_stats.total_xp,
            'level': self.user_stats.level,
            'xp_to_next_level': 100 - (self.user_stats.total_xp % 100),
            'total_sessions': self.user_stats.total_sessions,
            'current_streak': self.user_stats.current_streak,
            'best_streak': self.user_stats.best_streak,
            'achievements': self.user_stats.achievements,
            'weekly_sessions': self._get_recent_weekly_stats(),
            'monthly_sessions': self._get_recent_monthly_stats(),
        }

    def _get_recent_weekly_stats(self) -> List[Dict]:
        """Get last 8 weeks of session data"""
        weeks = []
        for i in range(7, -1, -1):  # Last 8 weeks including current
            week_date = date.today() - timedelta(weeks=i)
            week_key = self._get_week_key(week_date)
            sessions = self.user_stats.weekly_sessions.get(week_key, 0)
            weeks.append({
                'week': week_key,
                'sessions': sessions,
                'label': f"Week {week_date.isocalendar()[1]}"
            })
        return weeks

    def _get_recent_monthly_stats(self) -> List[Dict]:
        """Get last 6 months of session data"""
        months = []
        for i in range(5, -1, -1):  # Last 6 months including current
            month_date = date.today().replace(day=1) - timedelta(days=i*30)
            month_key = self._get_month_key(month_date)
            sessions = self.user_stats.monthly_sessions.get(month_key, 0)
            months.append({
                'month': month_key,
                'sessions': sessions,
                'label': month_date.strftime('%b %Y')
            })
        return months

    def reset_stats(self):
        """Reset all statistics (for testing purposes)"""
        self.user_stats = UserStats()
        self._save_stats()

File: static/js/gamification.js
// Gamification module for Pomodoro Timer
class GamificationUI {
    constructor() {
        this.currentTab = 'weekly';
        this.stats = null;
        this.init();
    }

    async init() {
        this.bindEvents();
        await this.loadStats();
        this.updateUI();
    }

    bindEvents() {
        // Stats tab switching
        document.querySelectorAll('.stats-tab').forEach(tab => {
            tab.addEventListener('click', (e) => {
                const tabName = e.target.dataset.tab;
                this.switchTab(tabName);
            });
        });

        // Modal close events
        document.getElementById('close-modal')?.addEventListener('click', () => {
            this.hideAchievementModal();
        });

        document.getElementById('close-levelup-modal')?.addEventListener('click', () => {
            this.hideLevelUpModal();
        });

        // Click outside modal to close
        document.getElementById('achievement-modal')?.addEventListener('click', (e) => {
            if (e.target.id === 'achievement-modal') {
                this.hideAchievementModal();
            }
        });

        document.getElementById('levelup-modal')?.addEventListener('click', (e) => {
            if (e.target.id === 'levelup-modal') {
                this.hideLevelUpModal();
            }
        });
    }

    async loadStats() {
        try {
            const response = await fetch('/api/stats');
            this.stats = await response.json();
        } catch (error) {
            console.error('Failed to load stats:', error);
            this.stats = {
                total_xp: 0,
                level: 1,
                xp_to_next_level: 100,
                total_sessions: 0,
                current_streak: 0,
                best_streak: 0,
                achievements: [],
                weekly_sessions: [],
                monthly_sessions: []
            };
        }
    }

    updateUI() {
        if (!this.stats) return;

        // Update level and XP
        document.getElementById('user-level').textContent = this.stats.level;
        document.getElementById('user-xp').textContent = `${this.stats.total_xp} XP`;
        
        // Update XP progress bar
        const xpProgress = ((100 - this.stats.xp_to_next_level) / 100) * 100;
        document.getElementById('xp-progress').style.width = `${xpProgress}%`;

        // Update streak
        document.getElementById('current-streak').textContent = this.stats.current_streak;

        // Update basic progress
        document.getElementById('sessions-completed').textContent = `${this.stats.total_sessions} Completed`;
        const focusTime = this.calculateFocusTime(this.stats.total_sessions);
        document.getElementById('focus-time').textContent = focusTime;

        // Update achievements
        this.updateAchievements();

        // Update statistics chart
        this.updateChart();
    }

    calculateFocusTime(sessions) {
        const totalMinutes = sessions * 25; // 25 minutes per session
        const hours = Math.floor(totalMinutes / 60);
        const minutes = totalMinutes % 60;
        return `${hours}h ${minutes}m`;
    }

    updateAchievements() {
        const grid = document.getElementById('achievements-grid');
        if (!grid) return;

        // Define all possible achievements
        const allAchievements = [
            { id: 'first_completion', name: 'First Steps', description: 'Complete your first Pomodoro', icon: '🍅' },
            { id: 'daily_streak_3', name: 'Consistent', description: '3 day streak', icon: '🔥' },
            { id: 'daily_streak_7', name: 'Dedicated', description: '7 day streak', icon: '⚡' },
            { id: 'weekly_10', name: 'Weekly Warrior', description: '10 sessions in a week', icon: '💪' },
            { id: 'weekly_20', name: 'Focus Master', description: '20 sessions in a week', icon: '👑' },
            { id: 'total_50', name: 'Half Century', description: '50 total sessions', icon: '🏆' },
            { id: 'total_100', name: 'Centurion', description: '100 total sessions', icon: '🌟' }
        ];

        const earnedIds = new Set(this.stats.achievements.map(ach => ach.id));

        grid.innerHTML = allAchievements.map(achievement => {
            const earned = earnedIds.has(achievement.id);
            return `
                <div class="achievement ${earned ? 'earned' : ''}" title="${achievement.description}">
                    <div class="achievement-icon">${achievement.icon}</div>
                    <div class="achievement-name">${achievement.name}</div>
                </div>
            `;
        }).join('');
    }

    updateChart() {
        const chartContainer = document.getElementById('stats-chart');
        if (!chartContainer) return;

        const data = this.currentTab === 'weekly' ? this.stats.weekly_sessions : this.stats.monthly_sessions;
        const maxSessions = Math.max(...data.map(item => item.sessions), 1);

        chartContainer.innerHTML = data.map(item => {
            const height = (item.sessions / maxSessions) * 80; // Max height 80px
            return `
                <div class="chart-bar" style="height: ${height}px;">
                    <div class="chart-value">${item.sessions}</div>
                    <div class="chart-label">${item.label}</div>
                </div>
            `;
        }).join('');
    }

    switchTab(tabName) {
        this.currentTab = tabName;
        
        // Update tab buttons
        document.querySelectorAll('.stats-tab').forEach(tab => {
            tab.classList.toggle('active', tab.dataset.tab === tabName);
        });

        // Update chart
        this.updateChart();
    }

    async onSessionComplete() {
        try {
            const response = await fetch('/api/complete-session', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                }
            });

            const result = await response.json();
            
            // Reload stats and update UI
            await this.loadStats();
            this.updateUI();

            // Show notifications
            if (result.level_up) {
                this.showLevelUpModal(result.level);
            }

            if (result.new_achievements && result.new_achievements.length > 0) {
                // Show achievement modal for first new achievement
                setTimeout(() => {
                    this.showAchievementModal(result.new_achievements[0]);
                }, result.level_up ? 2000 : 0); // Delay if there was a level up
            }

            return result;
        } catch (error) {
            console.error('Failed to complete session:', error);
            return null;
        }
    }

    showAchievementModal(achievement) {
        const modal = document.getElementById('achievement-modal');
        const icon = document.getElementById('modal-achievement-icon');
        const name = document.getElementById('modal-achievement-name');
        const desc = document.getElementById('modal-achievement-desc');

        if (modal && icon && name && desc) {
            icon.textContent = achievement.icon;
            name.textContent = achievement.name;
            desc.textContent = achievement.description;
            modal.classList.remove('hidden');
        }
    }

    hideAchievementModal() {
        const modal = document.getElementById('achievement-modal');
        if (modal) {
            modal.classList.add('hidden');
        }
    }

    showLevelUpModal(level) {
        const modal = document.getElementById('levelup-modal');
        const levelSpan = document.getElementById('modal-new-level');

        if (modal && levelSpan) {
            levelSpan.textContent = level;
            modal.classList.remove('hidden');
        }
    }

    hideLevelUpModal() {
        const modal = document.getElementById('levelup-modal');
        if (modal) {
            modal.classList.add('hidden');
        }
    }
}

// Initialize gamification system when DOM is loaded
document.addEventListener('DOMContentLoaded', () => {
    window.gamification = new GamificationUI();
});